gramps/gen/filters/optimizer.py

```diff
@@ -54,79 +54,53 @@ def union(sets: List[Set[PrimaryObjectHandle]]) -> Set[PrimaryObjectHandle]:
 
 
 class Optimizer:
-    """
-    Optimizer uses the filter's pre-selected selected_handles
-    for each rule to reduce the search space.
-    selected_handles is a superset of the final result of the rule.
-    the selected_handles of each rule in each filter of the filter list
-    are combined to create `possible_handles`, the superset of the result
-    of the filter list.
-    """
-
-    def __init__(self, all_handles: Set[PrimaryObjectHandle], filter: GenericFilter):
-        """
-        Compute possible_handles for the filter list.
-        """
-        self.all_handles = all_handles
-        self.possible_handles = self.compute_potential_handles_for_filter(filter)
-
     def compute_potential_handles_for_filter(
         self, filter: GenericFilter
     ) -> Set[PrimaryObjectHandle]:
-        """
-        Compute the superset of handles which are the result of the supplied filter
-        In the worst case, this is self.all_handles
-        """
         if len(filter.flist) == 0:
-            return self.all_handles
-        # for rules which are not optimized i.e. rules which do not have a
-        # `selected_handles` attribute
-        # compute the default set of handles which potentially match
-        # If the filter is inverted this is an empty set so that when, later,
-        # we compute the difference to self.all_handles, no handles are removed.
-        default_potential_handles = set() if filter.invert else self.all_handles
-        handlesets: List[Set[PrimaryObjectHandle]] = [
-            self.compute_potential_handles_for_rule(rule, default_potential_handles)
-            for rule in filter.flist
-        ]
-
-        if filter.logical_op == "and":
-            # the result of filter is contained within the intersection of the sets in handlesets
-            handles = intersection(handlesets)
-        elif filter.logical_op in ("or", "one"):
-            # the result of filter is contained within the union of the sets in handlesets
-            handles = union(handlesets)
+            return (None, None)
+
+        handles_in = None
+        handles_out = None
+        for rule in filter.flist:
+            rule_in, rule_out = self.compute_potential_handles_for_rule(rule)
+            if filter.logical_op == "and":
+                if rule_in is not None:
+                    if handles_in is None:
+                        handles_in = rule_in
+                    else:
+                        handles_in.intersection(rule_in)
+                if rule_out is not None:
+                    if handles_out is None:
+                        handles_out = rule_out
+                    else:
+                        handles_out.intersection(rule_out)
```

Comment:
Should this be union?
Given a filter with two rules and logical_op "and"
rule 1 says handle 'a' is out. So the path `handles_out = rule_out` is executed
rule 2 says handle 'b' is out. So the path `handles_out.intersection(rule_out)` is executed. This expands to 'a'.intersection('b') which is obviously an empty set

Logically, neither of handles 'a' or 'b' can be in the result so I would expect `handles_out` for the filter to be the set ('a', 'b')

Reply:
I think it will produce the correct result as it is, but `GenericFilter` will call `apply_logical_op` for handles 'a' and 'b' when it doesn't really need to

Reply:
> Should this be union?

I think intersection is correct, because the set can only get smaller.

Reply:
> I think it will produce the correct result as it is, but GenericFilter will call apply_logical_op for handles 'a' and 'b' when it doesn't really need to

Not sure I follow this. You want to make a suggestion?

Reply:
> This expands to 'a'.intersection('b') which is obviously an empty set

I think it is the same set. I think intersection is correct.

```ipython
In [1]: a = {1, 2, 3}
In [2]: b = {1, 2, 3}
In [3]: a.intersection(b)
Out[3]: {1, 2, 3}
```